batch/sql/estimated-current.sql

```diff
@@ -711,6 +719,22 @@ BEGIN
   END IF;
 END $$
 
+DROP TRIGGER IF EXISTS jobs_before_insert $$
+CREATE TRIGGER jobs_before_insert BEFORE INSERT ON jobs
+FOR EACH ROW
+BEGIN
+  DECLARE job_group_cancelled BOOLEAN;
+
+  SET job_group_cancelled = EXISTS (SELECT TRUE
+                                    FROM job_groups_cancelled
+                                    WHERE id = NEW.batch_id AND job_group_id = NEW.job_group_id
+                                    LOCK IN SHARE MODE);
+
+  IF job_group_cancelled THEN
+    SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = "job group has already been cancelled";
+  END IF;
+END $$
```

Comment:
The issue is a job group can be cancelled, but the batch isn't necessarily cancelled. Again, this could happen once we start adding new job groups. I can defer the change, but it's eventually needed. I was trying to get all of the remaining database changes done at once.

Reply:
Why can't the front end just also that the job group isn't cancelled?

Reply:
Sure. We could do that, but without adding a cache, we'd be making the same query over and over again. I thought this was simpler, but we could also get rid of that complexity for now. However, thinking about this more, what happens if a job group is cancelled in the middle of this transaction?

Reply:
I think this is the exact same problem that we currently have with batches (we query for the cancellation status of a batch on every operation that could update it like inserting jobs. I'd rather have two of the same problem than two different approaches. We could change them later on. Regarding the cancellation, that's why we select the batch for update before inserting jobs, so cancellation would only happen after the jobs are entered. I presume we could do the same for job groups.

Reply:
I think we need this trigger here. The issue is if we check whether the job group is cancelled when creating each job, we can't take a lock on the `job_groups_cancelled` table for that job group if it doesn't exist. I'm not sure why we don't have this same problem with batches. Maybe we do have the same issue and should fix it separately.

```python3
            record = await db.select_and_fetchone(
                '''
SELECT 1
FROM job_groups_cancelled
WHERE id = %s AND job_group_id = %s
LOCK IN SHARE MODE;
''',
                (batch_id, update_id, user),
            )
```